### ISSUE TITLE
Allow simultaneous login with multiple OAuth2 providers, with account linking

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1277]
**Is your feature request related to a problem? Please describe.**

I want to be able to have a user log in with Google, then also log in with Facebook (in order to pull photos from the Facebook account), and have both those login methods linked to the same Serverpod user.

Currently, as I understand it, you can only be logged in with one OAuth method at any given time in Serverpod.

**Describe the solution you'd like**

The login API should be extended to allow secondary logins, with linking to the primary login, if the user is already logged in on the primary account.

Also, currently Serverpod OAuth2 providers (notably the Google provider) logs the user out of OAuth2 after logging into ServerPod, but it would be useful to keep around a long-lived token that could be used to fetch data from the user's profile while they are logged in (e.g. photos).

**Comment by vlidholt** [comment]
Different sign-in types can be associated with a single user in Serverpod. Accounts are automatically linked based on email (if available).

**Comment by lukehutch** [comment]
Sorry, I didn't realize that already worked. However there is still an issue: currently OAuth login providers log the user out of the OAuth session as soon as Serverpod login is complete. This means the server and for example fetch an album of photos for the user later. I made a comment or two about long-lived tokens in my Facebook login implementation. But also, the code in the Google login provider that logs put the user should probably be removed, and the token should be saved in the database (in the Facebook provider, I do convert the token to a long-lived token, and store it in the database).

**Comment by vlidholt** [comment]
It sounds like we should give an option not to log out the user.

**Comment by lukehutch** [comment]
> It sounds like we should give an option not to log out the user.

Right, and if this option is enabled, then the short-lasting login token should be exchanged for a long-lasting token (most OAuth providers require this exchange).

**Comment by SandPod** [comment]
@lukehutch - I updated the title to better reflect the issue.

Please modify it again if you have a suggestion for a more accurate title.
